File: backend/src/main/java/org/example/musicstream/service/TrackService.java
package org.example.musicstream.service;

import org.example.musicstream.dto.TrackCreateDTO;
import org.example.musicstream.dto.TrackResponseDTO;
import org.example.musicstream.dto.TrackUpdateDTO;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.List;

public interface TrackService {

    TrackResponseDTO createTrack(TrackCreateDTO dto);

    TrackResponseDTO getTrackById(Long id);

    List<TrackResponseDTO> getAllTracks();

    TrackResponseDTO updateTrack(Long id, TrackUpdateDTO dto);

    TrackResponseDTO uploadAudioFile(Long id, MultipartFile file) throws IOException;

    void deleteTrack(Long id);
}

File: frontend/src/app/core/services/track.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Track, TrackCreateDTO, TrackUpdateDTO } from '../models/track.model';
import { environment } from '../../../environments/environment';

@Injectable({
    providedIn: 'root'
})
export class TrackService {
    private http = inject(HttpClient);
    private apiUrl = `${environment.apiUrl}/tracks`;

    getAllTracks(): Observable<Track[]> {
        return this.http.get<Track[]>(this.apiUrl);
    }

    getTrackById(id: number): Observable<Track> {
        return this.http.get<Track>(`${this.apiUrl}/${id}`);
    }

    createTrack(dto: TrackCreateDTO): Observable<Track> {
        return this.http.post<Track>(this.apiUrl, dto);
    }

    updateTrack(id: number, dto: TrackUpdateDTO): Observable<Track> {
        return this.http.put<Track>(`${this.apiUrl}/${id}`, dto);
    }

    deleteTrack(id: number): Observable<void> {
        return this.http.delete<void>(`${this.apiUrl}/${id}`);
    }

    uploadAudio(id: number, file: File): Observable<Track> {
        const formData = new FormData();
        formData.append('file', file);
        return this.http.post<Track>(`${this.apiUrl}/${id}/upload`, formData);
    }

    getStreamUrl(id: number): string {
        return `${this.apiUrl}/${id}/stream`;
    }
}
